### PR TITLE
STITCH-1953 Stop syncing documents on irrecoverable errors

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfigUnitTests.kt [hunk 1]
```diff
@@ -81,17 +81,28 @@ class CoreDocumentSynchronizationConfigUnitTests {
...
 
         config.isStale = false
+        config.setSomePendingWrites(
+            1,
+            ChangeEvent.changeEventForLocalInsert(
+                coll.namespace, BsonDocument("_id", BsonObjectId()), true))
+
         doc = config.toBsonDocument()
-        assert(!doc.getBoolean(CoreDocumentSynchronizationConfig.ConfigCodec.Fields.IS_STALE).value)
+        assert(
```
Comment (edaniels) [inline core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfigUnitTests.kt]: Just use assertTrue and assertFalse here.

File: server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 2]
```diff
@@ -808,6 +808,86 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
...
+                    errorEmitted = true
+                }
+            )
+
+            // insert an initial doc
+            val testDoc = Document("hello", "world")
+            val result = testSync.insertOneAndSync(testDoc)
+
+            // do a sync pass, synchronizing the doc
+            streamAndSync()
```
Comment (edaniels) [inline server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: [nit] After you stream and sync. Validate it's available remotely.

File: android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 3]
```diff
@@ -810,6 +813,86 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
...
+            // insert an initial doc
+            val testDoc = Document("hello", "world")
+            val result = Tasks.await(testSync.insertOneAndSync(testDoc))
+
+            // do a sync pass, synchronizing the doc
+            streamAndSync()
+
+            // update the doc
+            val expectedDoc = Document("hello", "computer")
+            testSync.updateOneById(result.insertedId, Document(Document("\$set", expectedDoc)))
```
Comment (edaniels) [inline android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: Await this

File: android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 4]
```diff
@@ -810,6 +813,86 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
...
+            // do a sync pass, and throw an error during the update
+            // freezing the document
+            streamAndSync()
+            Assert.assertTrue(errorEmitted)
+
+            // update the doc remotely
+            val nextDoc = Document("hello", "friend")
+
+            var sem = watchForEvents(namespace)
+            remoteColl.updateOne(Document("_id", result.insertedId), nextDoc)
```
Comment (edaniels) [inline android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: Await this

File: android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 5]
```diff
@@ -810,6 +813,86 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
...
+            sem.acquire()
+            streamAndSync()
+
+            // it should not have updated the local doc, as the local doc should be frozen
+            assertEquals(
+                withoutId(expectedDoc),
+                withoutVersionId(withoutId(Tasks.await(testSync.find(Document("_id", result.insertedId)).first())!!)))
+
+            // update the local doc. this should unfreeze the config
+            testSync.updateOneById(result.insertedId, Document("\$set", Document("no", "op")))
```
Comment (edaniels) [inline android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: Await this

File: android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 6]
```diff
@@ -810,6 +813,86 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
...
+            // this should still be the remote doc since remote wins
+            assertEquals(
+                withoutId(nextDoc),
+                withoutVersionId(withoutId(Tasks.await(testSync.find(Document("_id", result.insertedId)).first())!!)))
+
+            // update the doc remotely
+            val lastDoc = Document("good night", "computer")
+
+            sem = watchForEvents(namespace)
+            remoteColl.updateOne(
```
Comment (edaniels) [inline android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: Await this

File: android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 7]
```diff
@@ -810,6 +813,86 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
...
+                Document("_id", result.insertedId),
+                lastDoc
+            )
+            sem.acquire()
+
+            // now that we're sync'd and unfrozen, it should be reflected locally
+            // TODO: STITCH-1958 Possible race condition here for update listening
+            streamAndSync()
```
Comment (edaniels) [inline android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: Why is this two times?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java [hunk 8]
```diff
@@ -156,17 +158,44 @@ public void setStale(final boolean stale) {
...
+   * In the event that an irrecoverable error has occurred,
+   * updates should no longer be processed for a document.
+   * The config should reflect that it is in a frozen state.
+   *
+   * When a new write operation happens to this document locally
+   * we should unfreeze the document and resume trying to sync it.
+   *
+   * @param isFrozen whether or not this config is frozen
+   */
+  void setFrozen(final boolean isFrozen) {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: [major] In general setting anything must be reflected by taking a write lock out on the document and updating the collection with this document with frozen now being set. See setStale

**Comment by edaniels** [comment]
Needs some unit tests for frozen checks in data synchronizer but once done, LGTM.

**Comment by jsflax** [comment]
@edaniels PTAL

**Pull request merged**